### PR TITLE
BUG: Prevent recursive loop in some PDF files

### LINKED ISSUE #2474: BUG: clean_forms function cause infinite looping if elt["/Resources"] have circular relation
I have several PDF files from customers, and using remove_text can cause infinite looping. Upon investigation, I discovered a corner case where elt["/Resources"] has a circular relation, which can result in calling clean_forms(content, stack + [elt]) infinitely.

I proposed keeping a memory variable visited_resources to keep track of which elt["/Resources"] has been processed and to avoid infinite looping.

The corner case files are private and cannot be shared, but I believe many people would encounter the same problem.

### PR DESCRIPTION
**Pull request by pubpub-zz**
closes #2474
analysis in #2477

### CODE REVIEW

**Comment by pubpub-zz** [comment]
waiting for feed back from @syanng

**Comment by pubpub-zz** [comment]
@stefan6419846 
As coverage is not reduced, I propose to review and commit this PR.
@syanng please issue a new PR using your test file when available

File: pypdf/_writer.py [hunk 1]
```diff
@@ -1923,7 +1923,14 @@ def clean_forms(
             elt: DictionaryObject, stack: List[DictionaryObject]
         ) -> Tuple[List[str], List[str]]:
             nonlocal to_delete
-            if elt in stack:
+            # elt in recursive is a new contentstream object so we have to check the indirect_reference
```
Comment (stefan6419846) [inline pypdf/_writer.py:1926]:
```suggestion
            # elt in recursive call is a new ContentStream object, so we have to check the indirect_reference
```

File: pypdf/_writer.py [hunk 2]
```diff
@@ -1958,7 +1965,12 @@ def clean_forms(
                                     if k1 not in ["/Length", "/Filter", "/DecodeParms"]
                                 }
                             )
-                        clean_forms(content, stack + [elt])  # clean sub forms
+                            try:
+                                content.indirect_reference = o.indirect_reference
+                            except AttributeError:  # pragma: no cover
+                                pass
+                        stack.append(elt)
+                        clean_forms(content, stack)  # clean sub forms
```
Comment (stefan6419846) [inline pypdf/_writer.py:1973]:
```suggestion
                        clean_forms(content, stack)  # clean subforms
```

**Review by stefan6419846 — APPROVED** [review]

**Pull request merged**